src/music_gui.cpp

```diff
@@ -35,278 +36,388 @@
 
 #include "safeguards.h"
 
-/**
- * Get the name of the song.
- * @param index of the song.
- * @return the name of the song.
- */
-static const char *GetSongName(int index)
+
+struct MusicSystem {
+	struct PlaylistEntry : MusicSongInfo {
+		const MusicSet *set;  ///< music set the song comes from
+		int set_index;        ///< index of song in set
+
+		PlaylistEntry(const MusicSet *set, int set_index) : MusicSongInfo(set->songinfo[set_index]), set(set), set_index(set_index) { }
+		bool IsValid() const { return !StrEmpty(this->songname); }
+	};
+	using Playlist = std::vector<PlaylistEntry>;
+
+	enum PlaylistChoices {
+		PLCH_ALLMUSIC,
+		PLCH_OLDSTYLE,
+		PLCH_NEWSTYLE,
+		PLCH_EZYSTREET,
+		PLCH_CUSTOM1,
+		PLCH_CUSTOM2,
+		PLCH_THEMEONLY,
+		PLCH_MAX,
+	};
+
+	Playlist active_playlist;    ///< current play order of songs, including any shuffle
+	Playlist displayed_playlist; ///< current playlist as displayed in GUI, never in shuffled order
+	Playlist music_set;          ///< all songs in current music set, in set order
+
+	PlaylistChoices selected_playlist;
+
+	void BuildPlaylists();
+
+	void ChangePlaylist(PlaylistChoices pl);
+	void ChangeMusicSet(const std::string &set_name);
+	void Shuffle();
+	void Unshuffle();
+
+	void Play();
+	void Stop();
+	void Next();
+	void Prev();
+	void CheckStatus();
+
+	bool IsPlaying() const;
+	bool IsShuffle() const;
+	PlaylistEntry GetCurrentSong() const;
+
+	bool IsCustomPlaylist() const;
+	void PlaylistAdd(size_t song_index);
+	void PlaylistRemove(size_t song_index);
+	void PlaylistClear();
+
+private:
+	void ChangePlaylistPosition(int ofs);
+	int playlist_position;
+
+	void SaveCustomPlaylist(PlaylistChoices pl);
+
+	Playlist standard_playlists[PLCH_MAX];
+};
+
+MusicSystem _music;
+
+
+/** Rebuild all playlists for the current music set */
+void MusicSystem::BuildPlaylists()
 {
-	return BaseMusic::GetUsedSet()->songinfo[index].songname;
+	const MusicSet *set = BaseMusic::GetUsedSet();
+	int themenum = 0;
+
+	/* Clear current playlists */
+	for (auto &pl : this->standard_playlists) pl.clear();
```

Comment:
Not sure we really want to make the switch to C++11 right now.

Reply:
This probably needs a separate discussion, but is there a particular reason not to? The previous reason was the compile farm, which isn't an issue now. The build system needs a little bit of work to ensure compiler support for it, but other than that...

Reply:
Our compile farm is neither the only way for people to get binaries nor does all platforms people run OpenTTD on. Don't misunderstand me, I'm not generally against moving to C++11, but I am against sneaking it in without previous discussion in some "random" patch.